ASoC: SOF: Intel: hda: Add support for persistent Code Loader DMA buffers

ujfalusi opened the pull request:
Hi,

every now and then we see the DMA buffer allocation failure coming back [1]. While the issue is not originating from SOF stack, it does pops up from the logs.

It has been reported that the DMA memory allocation for firmware download
can fail after extended period of uptime on systems with relatively small
amount of RAM when the system memory becomes fragmented.

The issue primarily happens  when the system is waking up from system
suspend, swap might not be available and the MM system cannot move things
around to allow for successful allocation.

If the IMR boot is not supported then for each DSP boot we would need to
allocate the DMA buffer for firmware transfer, which can increase the
chances of the issue to be hit.

This patch adds support for allocating the DMA buffers once at first boot
time and keep it until the system is shut down, rebooted or the drivers
re-loaded and makes this as the default operation.

With persistent_cl_buffer module parameter the persistent Code Loader
DMA buffer can be disabled to fall back to on demand allocation.

[1] https://github.com/thesofproject/linux/issues/5161

ujfalusi committed:
ASoC: SOF: Intel: hda: Add support for persistent Code Loader DMA buffers

It has been reported that the DMA memory allocation for firmware download
can fail after extended period of uptime on systems with relatively small
amount of RAM when the system memory becomes fragmented.

The issue primarily happens  when the system is waking up from system
suspend, swap might not be available and the MM system cannot move things
around to allow for successful allocation.

If the IMR boot is not supported then for each DSP boot we would need to
allocate the DMA buffer for firmware transfer, which can increase the
chances of the issue to be hit.

This patch adds support for allocating the DMA buffers once at first boot
time and keep it until the system is shut down, rebooted or the drivers
re-loaded and makes this as the default operation.

With persistent_cl_buffer module parameter the persistent Code Loader
DMA buffer can be disabled to fall back to on demand allocation.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

plbossart commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -252,10 +252,9 @@ int hda_cl_trigger(struct device *dev, struct hdac_ext_stream *hext_stream, int
-int hda_cl_cleanup(struct device *dev, struct snd_dma_buffer *dmab,
-		   struct hdac_ext_stream *hext_stream)
+static int hda_cl_cleanup(struct device *dev, struct snd_dma_buffer *dmab,
+			  struct hdac_ext_stream *hext_stream)
```
this was exported for a reason, it'll be needed in the BPT/BRA stuff, see https://github.com/thesofproject/linux/pull/4679/commits/af779a3ca9bae844fb1691caaba7243e49afbe71

If you remove the export, you'll have to add it back at some point.

ujfalusi replied:
OK, dropping this patch

plbossart commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -286,11 +285,11 @@ int hda_cl_cleanup(struct device *dev, struct snd_dma_buffer *dmab,
 
-int hda_cl_copy_fw(struct snd_sof_dev *sdev, struct hdac_ext_stream *hext_stream)
+static int hda_cl_copy_fw(struct snd_sof_dev *sdev,
+			  struct hdac_ext_stream *hext_stream)
```
that one is probably fine to remove, I don't recall that it's used by BPT/BRA.

ujfalusi replied:
I will just drop this as it is.

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -383,6 +383,20 @@ static inline bool hda_sdw_check_wakeen_irq(struct snd_sof_dev *sdev)
+
+		/* Check if IMR boot is usable */
+		if (!sof_debug_check_flag(SOF_DBG_IGNORE_D3_PERSISTENT) &&
+		    (sdev->fw_ready.flags & SOF_IPC_INFO_D3_PERSISTENT ||
```
are the fw_ready flags set before the pre_fw_run callback? This would need to be commented with more details.

ujfalusi replied:
aargh, you are right. The fw_ready.flags are coming from the FW_READY message _after_ the fw have booted.

If I want to keep the 'clever' way to keep the DMA buffer if IMR boot is not supported then we would only be able to store the buffer on the second boot. That could be fine, but it is a bit clumsy to document.

Probably dropping the cleverness and add a Kconfig instead to force this for products that would benefit.

plbossart commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -61,11 +62,14 @@ hda_cl_prepare(struct device *dev, unsigned int format, unsigned int size,
-	if (ret < 0) {
-		dev_err(sdev->dev, "error: memory alloc failed: %d\n", ret);
-		goto out_put;
+	if (!keep_buffer || (keep_buffer && !dmab->area)) {
```
this sort of test gives me a migraine.

ujfalusi replied:
We need to allocate the DMA buffer if we don't want to keep the DMA buffer OR if we want to keep the buffer, but it has not yet been allocated.
I can add a comment.

plbossart commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -278,10 +282,12 @@ static int hda_cl_cleanup(struct device *dev, struct snd_dma_buffer *dmab,
-	dmab->area = NULL;
-	hstream->bufsize = 0;
-	hstream->format_val = 0;
+	if (!keep_buffer) {
```
that's not symmetrical with 

````c
if (!keep_buffer || (keep_buffer && !dmab->area)) {
````

so it's anyone's guess when then memory is freed when the second part of the OR was true.

ujfalusi replied:
It does not need to be as hda_cl_cleanup cannot check for !dmab->area since if we are here the buffer has been already allocated, it is just a matter of do we want to keep it or not.
I can add a comment here as well.

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -395,6 +400,11 @@ int hda_dsp_pre_fw_run(struct snd_sof_dev *sdev)
+		if (hda_keep_fw_dma_buffer == 0 || hda_keep_fw_dma_buffer == 1)
+			hdev->keep_fw_dma_buffer = hda_keep_fw_dma_buffer;
+		else
+			hdev->keep_fw_dma_buffer = !hdev->imrboot_supported;
```
this is become hard to track, IIRC there was a case where if the firmware crashes in the previous run then IMR is disabled and we would need a DMA buffer.

Why not just allocate the DMA buffer always, even if IMR is used after the first boot.

ujfalusi replied:
You mean keeping the DMA buffer unconditionally?

This does not change how we do IMR, it is only aimed to give an option that we can do the DMA buffer allocation and setup only once and if needed use it.

We can make this opt-out, so that by default we would keep the buffer and users can opt out?

plbossart commented on sound/soc/sof/intel/hda.h:
```diff
@@ -489,12 +489,20 @@ enum sof_hda_D0_substate {
 
+	/*
+	 * DMA buffer for base firmware download. If keep_fw_dma_buffer is true,
+	 * the buffer is allocated once and kept through the liftime of the
```
typo: lifetime

ujfalusi commented:
Changes since v1:
- single patch
- move the module parameter to hda-loader.c and use it directly
- the firmware download DMA buffer is by default persistent
- add explanatory comments

plbossart commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -279,10 +296,13 @@ int hda_cl_cleanup(struct device *dev, struct snd_dma_buffer *dmab,
-	hstream->bufsize = 0;
-	hstream->format_val = 0;
+
+	if (!keep_buffer) {
```
nit-pick: use same working instead of having 'keep_buffer' and 'persistent_buffer'. If it's the same thing use the same name, otherwise readers might think they are different concepts.

ujfalusi replied:
OK, I'll use `persistent_buffer` in both function, I think the kernel parameter name can remain as it is?

plbossart replied:
I would use the same naming everywhere but having the variables named as 'persistent' is good enough,

ujfalusi commented:
Changes since v2:
- Use persistent_buffer as parameter names in functions

plbossart commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -558,8 +585,8 @@ int hda_dsp_ipc4_load_library(struct snd_sof_dev *sdev,
 	hext_stream = hda_cl_prepare(sdev->dev, HDA_CL_STREAM_FORMAT,
-				     stripped_firmware.size,
-				     &dmab, SNDRV_PCM_STREAM_PLAYBACK, false);
+				     stripped_firmware.size, &dmab, false,
```
don't we also need to keep this buffer permanently? 

Not sure I understand why the firmware buffer would be kept but the library would be treated differently.

ujfalusi replied:
The library (and iccmax capture?) buffer is not yet handled, the issues so far only been hit with older platforms running ipc3, hence no libraries nor iccmax.
iccmax should be relatively easy while the library is not as we don't know how many of them we have..

plbossart replied:
@ujfalusi are you saying that we need N buffers for N libraries? That would obviously be problematic.

One would think there's a way to reuse a single worst-case buffer for all this, at the end of the day the sum of base firmware + N libraries cannot exceed the size of the internal memory, can it?

ujfalusi replied:
@plbossart, hrm, right, I guess we could try to get 'big enough' buffer and just do memcpy to it.
Allocate for basefw first and if any of the library needs more space then re-allocate a bigger buffer and keep using that.
Probably I would have arrived to this in few days ;)

marc-hb commented:
I very painfully found this error buried deep down in the Jenkins logs. See internal issue 607 for more details.

It's failing only the NOCODEC build, the regular build is fine.

Yet Jenkins thinks the build was fine, keeps going and that's why things get super confusing.

Considering Jenkins is basically orphaned, you want to add the NODEC build to Github Actions. It shouldn't be hard, I can review if you want.

```
  CC      net/ethtool/rss.o
  CC      drivers/pnp/support.o
  CC      drivers/acpi/acpica/rsdumpinfo.o
/srv/home/jenkins/workspace/linux_config_build/linux/sound/soc/sof/intel/hda.c: In function â€˜hda_dsp_pre_fw_runâ€™:
/srv/home/jenkins/workspace/linux_config_build/linux/sound/soc/sof/intel/hda.c:404:21: error: â€˜hda_keep_fw_dma_bufferâ€™ undeclared (first use in this function)
  404 |                 if (hda_keep_fw_dma_buffer == 0 || hda_keep_fw_dma_buffer == 1)
      |                     ^~~~~~~~~~~~~~~~~~~~~~
/srv/home/jenkins/workspace/linux_config_build/linux/sound/soc/sof/intel/hda.c:404:21: note: each undeclared identifier is reported only once for each function it appears in
make[10]: *** [/srv/home/jenkins/workspace/linux_config_build/linux/scripts/Makefile.build:244: sound/soc/sof/intel/hda.o] Error 1
make[9]: *** [/srv/home/jenkins/workspace/linux_config_build/linux/scripts/Makefile.build:485: sound/soc/sof/intel] Error 2
make[9]: *** Waiting for unfinished jobs....
  AR      drivers/amba/built-in.a
  CC      drivers/pnp/interface.o
  CC      drivers/acpi/acpica/rsinfo.o
```

ujfalusi commented:
SOFCI TEST

marc-hb commented:
> Considering Jenkins is basically orphaned, you want to add the NODEC build to Github Actions. It shouldn't be hard, I can review if you want.

BTW you should really use a `matrix` to avoid the massive copy/paste/diverge in `.github/workflows/`. Simple matrix example in: https://github.com/thesofproject/sof/commit/425a5b55777f614a7e3210b75c1ba87f17717013

marc-hb commented:
SOFCI TEST

ujfalusi commented:
Changes since v3:
- Preserve the iccmax DMA buffer if it is used by the platform
- With this update the whole base firmware handling is covered, IPC4 library DMA buffer caching will be added as a followup series as it needs to touch top-level IPC4 headers and structures - only Intel is using `struct snd_dma_buffer` for download afaik.

ujfalusi commented:
Changes since v4:
- cover library loading as well with persistent code loader buffer
- single cl_dmab buffer to be used by basefw and library loading alike.

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -936,6 +936,12 @@ void hda_dsp_remove(struct snd_sof_dev *sdev)
+	if (hda->cl_dmab.area)
+		snd_dma_free_pages(&hda->cl_dmab);
+	if (hda->iccmax_dmab.area)
+		snd_dma_free_pages(&hda->iccmax_dmab);
```
I don't fully remember the iccmax details but would it be possible to use the same trick and use the same buffer as for regular firmware?

ujfalusi replied:
I rather not mix the two, while iccmax afaik is not running a capture, the buffer needs to be valid before we set up the code loader DMA.
I think it would result rather confusing code to follow what is going on.

ranj063 commented on sound/soc/sof/intel/hda.h:
```diff
@@ -495,6 +495,15 @@ struct sof_intel_hda_dev {
+	/*
+	 * DMA buffers for base firmware download. By default the buffers are
+	 * allocated once and kept through the lifetime of the driver.
+	 * See module parameter: keep_firmware_dma_buffer
```
where is this module parameter defined?

ujfalusi replied:
good catch, I renamed the parameter at the last time and missed this comment.

ranj063 commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -432,23 +455,31 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
+	 * Copy the payload to the DMA buffer if it is temporary or if the
+	 * buffer is  persistent but it does not have the basefw payload either
+	 * because this is the first time and needs to be initialized or a
+	 * library got loaded since the last basefw boot.
```
this comment is not easy to follow. What are we trying to say here? Also, if I understand the patch correctly, we're trying to propose to only keep the code loader DMA buffer allocation be pesistent isnt it? Why not then use the temp dmab in the iccmax and the library cases so that we never use the same buffer in those cases?

ujfalusi replied:
We are using the cl_dmab for basefw and library.
For iccmax I rather not use the same buffer by principle, iccmax is capture stream, code loader is playback for one.
We allocate iccmax stream first (with the buffer) and then the buffer for code loader, for iccmax we use one page, which rarely enough for the firmware, so we would need to free and allocate a new buffer, thus re-configure iccmax, but this will bring up other issues.

If we overlook that we would need to allocate by-directional DMA memory and absolutely be certain that iccmax is not capturing data (which it should not afaik) then we still need quite a big refactoring to do just to flip over the allocation flow for basefw.

For me the use the same buffer for playback and capture which sounds off limits.

plbossart replied:
Agree with Peter, the iccmax thing is a work-around only needed on a small set of platforms, it's best to keep a different buffer for iccmax capture and avoid mixing different types of uses.

ujfalusi commented:
Changes since v5:
- comments updated (module parameter rename and clarification)

bardliao commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -556,16 +587,28 @@ int hda_dsp_ipc4_load_library(struct snd_sof_dev *sdev,
+	if (persistent_cl_buffer && stripped_firmware.size > hda->cl_dmab.bytes) {
+		snd_dma_free_pages(&hda->cl_dmab);
+		hda->cl_dmab.area = NULL;
+		hda->cl_dmab.bytes = 0;
```
When will cl_dmab be allocated?

ujfalusi replied:
hda_cl_prepare() allocates the buffer if it is not yet allocated.
If the buffer which we allocated for the basefw is smaller than what we would need for a library, we free the allocated buffer and let hda_cl_prepare() to re-allocate a new, bigger buffer - which will be used for fw download as well later.
